internal/infrastructure/kubernetes/service.go

```diff
@@ -107,15 +107,98 @@ func (i *Infra) createOrUpdateService(ctx context.Context, infra *ir.Infra) erro
 	return nil
 }
 
-// deleteService deletes the Envoy Service in the kube api server, if it exists.
-func (i *Infra) deleteService(ctx context.Context, infra *ir.Infra) error {
+// deleteProxyService deletes the Envoy Service in the kube api server, if it exists.
+func (i *Infra) deleteProxyService(ctx context.Context, infra *ir.Infra) error {
 	svc := &corev1.Service{
 		ObjectMeta: metav1.ObjectMeta{
 			Namespace: i.Namespace,
-			Name:      expectedServiceName(infra.Proxy.Name),
+			Name:      expectedProxyServiceName(infra.Proxy.Name),
 		},
 	}
 
+	return i.deleteService(ctx, svc)
+}
+
+// expectedRateLimitInfraService returns the expected rate limit Service based on the provided infra.
+func (i *Infra) expectedRateLimitService(_ *ir.RateLimitInfra) *corev1.Service {
+	ports := []corev1.ServicePort{
+		{
+			Name:       "http",
+			Protocol:   corev1.ProtocolTCP,
+			Port:       rateLimitInfraHTTPPort,
+			TargetPort: intstr.IntOrString{IntVal: rateLimitInfraHTTPPort},
+		},
+	}
+
+	labels := rateLimitLabels()
+
+	svc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: i.Namespace,
+			Name:      rateLimitInfraName,
+			Labels:    labels,
+		},
+		Spec: corev1.ServiceSpec{
+			Type:            corev1.ServiceTypeLoadBalancer,
+			Ports:           ports,
+			Selector:        getSelector(labels).MatchLabels,
+			SessionAffinity: corev1.ServiceAffinityNone,
+			// Preserve the client source IP and avoid a second hop for LoadBalancer.
+			ExternalTrafficPolicy: corev1.ServiceExternalTrafficPolicyTypeLocal,
+		},
+	}
+
+	return svc
+}
+
+// createOrUpdateRateLimitService creates a Service in the kube api server based on the provided infra,
+// if it doesn't exist or updates it if it does.
+func (i *Infra) createOrUpdateRateLimitService(ctx context.Context, infra *ir.RateLimitInfra) error {
+	svc := i.expectedRateLimitService(infra)
+	return i.createOrUpdateService(ctx, svc)
+}
+
+// deleteRateLimitService deletes the rate limit Service in the kube api server, if it exists.
+func (i *Infra) deleteRateLimitService(ctx context.Context, _ *ir.RateLimitInfra) error {
+	svc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: i.Namespace,
+			Name:      rateLimitInfraName,
+		},
+	}
+
+	return i.deleteService(ctx, svc)
+}
+
+func (i *Infra) createOrUpdateService(ctx context.Context, svc *corev1.Service) error {
+	current := &corev1.Service{}
+	key := types.NamespacedName{
+		Namespace: svc.Namespace,
+		Name:      svc.Name,
+	}
+
+	if err := i.Client.Get(ctx, key, current); err != nil {
```

Comment:
can we make `L180-L197` reusable?

Reply:
looks like we'd move the complexity over to the util func -  we could  pass objects as`Client.Object` to it, which is fine, but we also do an equality check on `obj.Spec`, so we'd have to typecast it back to the orig obj type

Reply:
address and refactor this later